frame/elections-phragmen/src/lib.rs

```diff
@@ -350,53 +408,83 @@ decl_module! {
 		/// State reads:
 		/// 	- Candidates.len() + Members.len() + RunnersUp.len()
 		/// 	- Voting (is_voter)
+		/// 	- Lock
 		/// 	- [AccountBalance(who) (unreserve + total_balance)]
 		/// State writes:
 		/// 	- Voting
 		/// 	- Lock
 		/// 	- [AccountBalance(who) (unreserve -- only when creating a new voter)]
 		/// # </weight>
-		#[weight = 50 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(4, 2)]
+		#[weight = T::WeightInfo::vote(votes.len() as u32)]
 		fn vote(
 			origin,
 			votes: Vec<T::AccountId>,
 			#[compact] value: BalanceOf<T>,
-		) {
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			// votes should not be empty and more than `MAXIMUM_VOTE` in any case.
 			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
 			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
 
 			let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0);
 			let members_count = <Members<T>>::decode_len().unwrap_or(0);
 			let runners_up_count = <RunnersUp<T>>::decode_len().unwrap_or(0);
 
+			// can never submit a vote of there are no members, and cannot submit more votes than
+			// all potential vote targets.
 			// addition is valid: candidates, members and runners-up will never overlap.
 			let allowed_votes = candidates_count + members_count + runners_up_count;
-
 			ensure!(!allowed_votes.is_zero(), Error::<T>::UnableToVote);
 			ensure!(votes.len() <= allowed_votes, Error::<T>::TooManyVotes);
 
 			ensure!(value > T::Currency::minimum_balance(), Error::<T>::LowBalance);
 
-			// first time voter. Reserve bond.
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
+			// Reserve bond.
+			let Voter { votes: old_votes, deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			let maybe_refund = if old_votes.is_empty() {
+				// First time voter. Get full deposit.
+				let deposit = Self::deposit_of(votes.len());
+				T::Currency::reserve(&who, deposit)
 					.map_err(|_| Error::<T>::UnableToPayBond)?;
-			}
+				None
+			} else {
+				// Old voter.
+				match votes.len().cmp(&old_votes.len()) {
```

Comment:
This is bad.

Instead of messing with `len`s, just calculate the new amount to be reserved, and find the difference with what has currently been reserved. If it's greater, reserve the difference. If it's less, unreserve the differrence.

This way even if the calculation/constants for `deposit_factor_of` change between `vote`s, you'll still ensure that the latest prices are the ones that the user pays.

If you don't do that, then suppose the prices go up 2x between votes: A user could make 10 votes before, replace them with 5 of them later, be fully unreserved, yet still have 5 votes.

Reply:
right this logic was made with the assumption of _not_ storing the deposit.

Reply:
Addressed, thanks!